### ISSUE TITLE
JetBrains Tree causes NPE when rendering for XPath

### ISSUE THREAD

**Issue by abrooksv** [JetBrains/intellij-ui-test-robot#12]
I do not see this error in 0.9.35 and I do not have the source for that revision to help triage:

It looks to be the clipping rectangle is not set which Tree always expects in its paint method.
```
java.lang.NullPointerException
at com.intellij.ui.treeStructure.Tree.paint(Tree.java:219)
at com.intellij.remoterobot.services.dataExtractor.TextParser.parseData(TextParser.kt:36)
at com.intellij.remoterobot.services.dataExtractor.TextParser.parseComponent(TextParser.kt:22)
at com.intellij.remoterobot.services.xpath.XpathDataModelCreator$Companion.fillElement(XpathDataModelCreator.kt:137)
at com.intellij.remoterobot.services.xpath.XpathDataModelCreator$Companion.createElement(XpathDataModelCreator.kt:57)
at com.intellij.remoterobot.services.xpath.XpathDataModelCreator$Companion.addComponent(XpathDataModelCreator.kt:32)
at com.intellij.remoterobot.services.xpath.XpathDataModelCreator$Companion.addComponent(XpathDataModelCreator.kt:49)
at com.intellij.remoterobot.services.xpath.XpathDataModelCreator$Companion.addComponent(XpathDataModelCreator.kt:49)
at com.intellij.remoterobot.services.xpath.XpathDataModelCreator$Companion.addComponent(XpathDataModelCreator.kt:49)
at com.intellij.remoterobot.services.xpath.XpathDataModelCreator$Companion.addComponent(XpathDataModelCreator.kt:49)
at com.intellij.remoterobot.services.xpath.XpathDataModelCreator$Companion.addComponent(XpathDataModelCreator.kt:49)
at com.intellij.remoterobot.services.xpath.XpathDataModelCreator$Companion.access$addComponent(XpathDataModelCreator.kt:24)
at com.intellij.remoterobot.services.xpath.XpathDataModelCreator$create$1.executeInEDT(XpathDataModelCreator.kt:260)
at org.assertj.swing.edt.GuiTask.run(GuiTask.java:36)
at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:776)
at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
at java.base/java.security.AccessController.doPrivileged(Native Method)
at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:746)
at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:974)
at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:847)
at com.intellij.ide.IdeEventQueue.lambda$null$8(IdeEventQueue.java:449)
at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:739)
at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$9(IdeEventQueue.java:448)
at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:831)
at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:502)
at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
java.lang.IllegalStateException: Text parsing error. Can't do paint on Tree
at com.intellij.remoterobot.services.dataExtractor.TextParser$parseData$1.executeInEDT(TextParser.kt:43)
at org.assertj.swing.edt.GuiTask.run(GuiTask.java:36)
at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:776)
at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
at java.base/java.security.AccessController.doPrivile
```

**Comment by abrooksv** [comment]
I do not see the error when triggered from the browser view, but findAll() does trigger it

**Comment by abrooksv** [comment]
Tracing and debugging this more it seems to happen only when a Tree has a selection which causes the `TextParser#findContainerComponent` to return the Tree itself and not its `JBViewport` that is surrounding it. If I force the parent component to become the `JBViewport`, it works correctly.

**Comment by abrooksv** [comment]
Tracing it appears to be only when a Tree has a selection since that is what triggers the clipRect call.

The project view tree is not affected because its parent component that is rendered is the JBviewPort:

```
Component: Project com.intellij.ide.projectView.impl.ProjectViewPane$1[,0,0,767x399,alignmentX=0.0,alignmentY=0.0,border=com.intellij.ui.components.JBViewport$ResourceViewBorder@4dd851cc,flags=50331936,maximumSize=,minimumSize=,preferredSize=,editable=false,invokesStopCellEditing=false,largeModel=true,rootVisible=false,rowHeight=20,scrollsOnExpand=false,showsRootHandles=true,toggleClickCount=2,visibleRowCount=20]

containerComponent: com.intellij.ui.components.JBViewport[,0,0,440x399,layout=com.intellij.ui.components.JBViewport$1,alignmentX=0.0,alignmentY=0.0,border=,flags=25165832,maximumSize=,minimumSize=,preferredSize=,isViewSizeSet=true,lastPaintPosition=java.awt.Point[x=0,y=0],scrollUnderway=false]
```

Which makes sense since that is the containing element:
<img width="491" alt="Screen Shot 2021-01-05 at 4 20 16 PM" src="https://user-images.githubusercontent.com/8992246/103713709-e9f55400-4f71-11eb-9deb-cdd04698b528.png">

But my Tree's parentComponent is always itself:

```
Component: com.intellij.ui.treeStructure.Tree[,0,0,440x385,alignmentX=0.0,alignmentY=0.0,border=com.intellij.ui.components.JBViewport$ResourceViewBorder@f3f100d,flags=50332000,maximumSize=,minimumSize=,preferredSize=,editable=false,invokesStopCellEditing=false,largeModel=false,rootVisible=false,rowHeight=20,scrollsOnExpand=false,showsRootHandles=true,toggleClickCount=2,visibleRowCount=20]

containerComponent: com.intellij.ui.treeStructure.Tree[,0,0,440x385,alignmentX=0.0,alignmentY=0.0,border=com.intellij.ui.components.JBViewport$ResourceViewBorder@f3f100d,flags=50332000,maximumSize=,minimumSize=,preferredSize=,editable=false,invokesStopCellEditing=false,largeModel=false,rootVisible=false,rowHeight=20,scrollsOnExpand=false,showsRootHandles=true,toggleClickCount=2,visibleRowCount=20]
```

event though it has a surrounding JBViewport as well:
<img width="486" alt="Screen Shot 2021-01-05 at 4 21 45 PM" src="https://user-images.githubusercontent.com/8992246/103713842-1f9a3d00-4f72-11eb-8ec7-4573c07d7b0a.png">

---

If you shrink the tree horizontally enough so that the elements in the tree are clipped, it will work.
or
Stretch the project view enough that its contents fits fully in the tool window, then getting the text out of the project view tree will fail to work.

I have narrowed the issue down to `TextParser#findContainerComponent`'s width and height logic but I am not 100% sure what the purpose of the logic is so not sure what correct fix is.

Adding the following to the start of the method does make it appear to work though:
```
        if (component.parent is JViewport) {
            return component.parent
        }
```

**Comment by nizienko** [comment]
Indeed we've found that some components throws NPE when painting. And all cases were connected with parent which is smaller than itself. Here we decided to paint such parents instead of component except dialogs(dialog could be larger than some parents like welcome frame). What we've missed is that the component could fit the parent and still throws NPE. 
Seems that our assumption was not fully right. 
I'll replace this logic with yours checking if parents type is JViewport and will try our tests with it. Thanks for the investigation!
